src/main.c

```diff
@@ -0,0 +1,441 @@
+#include <unistd.h>
+
+struct nvme_lbaf
+{
+    __le16 ms;
+    __u8 ds;
+    __u8 rp;
+};
+
+struct nvme_id_ns
+{
+    __le64 nsze;
+    __le64 ncap;
+    __le64 nuse;
+    __u8 nsfeat;
+    __u8 nlbaf;
+    __u8 flbas;
+    __u8 mc;
+    __u8 dpc;
+    __u8 dps;
+    __u8 nmic;
+    __u8 rescap;
+    __u8 fpi;
+    __u8 dlfeat;
+    __le16 nawun;
+    __le16 nawupf;
+    __le16 nacwu;
+    __le16 nabsn;
+    __le16 nabo;
+    __le16 nabspf;
+    __le16 noiob;
+    __u8 nvmcap[16];
+    __le16 npwg;
+    __le16 npwa;
+    __le16 npdg;
+    __le16 npda;
+    __le16 nows;
+    __u8 rsvd74[18];
+    __le32 anagrpid;
+    __u8 rsvd96[3];
+    __u8 nsattr;
+    __le16 nvmsetid;
+    __le16 endgid;
+    __u8 nguid[16];
+    __u8 eui64[8];
+    struct nvme_lbaf lbaf[16];
+    __u8 rsvd192[192];
+    __u8 vs[3712];
+};
+
+#define MAX_PATH 4096
+#define MICROSOFT_NVME_VENDOR_ID 0x1414
+
+struct nvme_controller
+{
+    char name[256];
+    char dev_path[MAX_PATH];
+    char sys_path[MAX_PATH];
+    char model[MAX_PATH];
+};
+
+static bool debug = false;
+static bool udev_mode = false;
+
+#define DEBUG_PRINTF(fmt, ...)                                                                                         \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        if (debug == true)                                                                                             \
+        {                                                                                                              \
+            fprintf(stderr, "DEBUG: " fmt, ##__VA_ARGS__);                                                             \
+        }                                                                                                              \
+    } while (0)
+
+/**
+ * Given the path to a namespace, determine the namespace id.
+ */
+int get_namespace_id_for_path(const char *namespace_path)
+{
+    unsigned int ctrl, nsid;
+
+    if (sscanf(namespace_path, "/dev/nvme%un%u", &ctrl, &nsid) != 2)
+    {
+        return -1;
+    }
+
+    return nsid;
+}
+
+/**
+ * Dump environment variables.
+ */
+void debug_environment_variables()
+{
+    int i = 0;
+
+    DEBUG_PRINTF("Environment Variables:\n");
+    while (environ[i])
+    {
+        DEBUG_PRINTF("%s\n", environ[i]);
+        i++;
+    }
+}
+
+/**
+ * Read file contents as a null-terminated string.
+ *
+ * For sysfs entries, etc. where the file contents are not null-terminated and
+ * we know the contents are just a simple string (e.g. sysfs attributes).
+ */
+char *read_file_as_string(const char *path)
+{
+    DEBUG_PRINTF("reading %s...\n", path);
+
+    FILE *file = fopen(path, "r");
+    if (file == NULL)
+    {
+        perror("fopen");
+        return NULL;
+    }
+
+    // Determine the file size
```

Comment:
max sysfs attribute is like 4KB. Perhaps we can just allocate a 4KB block and read it without having to count the string?

Another option is to have the function take in a pre-allocated array and its max length, caller can just do something like this:
char buffer[MAX_SYSFS_LENGTH];
len = read_file_as_string(buffer, path); //len < 0 indicates error

Reply:
Are the tradeoffs worth it?  Maybe something to consider in future PR?

Reply:
I'm ok with re-evaluating this in the future.